conf/schemas.go

```diff
@@ -58,50 +91,114 @@ func ReadSchemas(file string) (Schemas, error) {
 		}
 
 		if sec.ValueOf("pattern") == "" {
-			return nil, fmt.Errorf("[%s]: empty pattern", schema.Name)
+			return Schemas{}, fmt.Errorf("[%s]: empty pattern", schema.Name)
 		}
 		schema.Pattern, err = regexp.Compile(sec.ValueOf("pattern"))
 		if err != nil {
-			return nil, fmt.Errorf("[%s]: failed to parse pattern %q: %s", schema.Name, sec.ValueOf("pattern"), err.Error())
+			return Schemas{}, fmt.Errorf("[%s]: failed to parse pattern %q: %s", schema.Name, sec.ValueOf("pattern"), err.Error())
 		}
 
 		schema.Retentions, err = ParseRetentions(sec.ValueOf("retentions"))
 		if err != nil {
-			return nil, fmt.Errorf("[%s]: failed to parse retentions %q: %s", schema.Name, sec.ValueOf("retentions"), err.Error())
+			return Schemas{}, fmt.Errorf("[%s]: failed to parse retentions %q: %s", schema.Name, sec.ValueOf("retentions"), err.Error())
 		}
 
 		p := int64(0)
 		if sec.ValueOf("priority") != "" {
 			p, err = strconv.ParseInt(sec.ValueOf("priority"), 10, 0)
 			if err != nil {
-				return nil, fmt.Errorf("Failed to parse priority %q for [%s]: %s", sec.ValueOf("priority"), schema.Name, err)
+				return Schemas{}, fmt.Errorf("Failed to parse priority %q for [%s]: %s", sec.ValueOf("priority"), schema.Name, err)
 			}
 		}
 		schema.Priority = int64(p)<<32 - int64(i) // to sort records with same priority by position in file
 
 		schemas = append(schemas, schema)
 	}
 
-	sort.Sort(schemas)
-	return schemas, nil
+	sort.Sort(SchemaSlice(schemas))
+
+	return NewSchemas(schemas), nil
 }
 
 // Match returns the correct schema setting for the given metric
 // it can always find a valid setting, because there's a default catch all
 // also returns the index of the setting, to efficiently reference it
-func (s Schemas) Match(metric string) (uint16, Schema) {
-	for i, schema := range s {
+func (s Schemas) Match(metric string, interval int) (uint16, Schema) {
```

Comment:
I agree that it's kind of hard to wrap ones head around that concept, but it's pretty cool that in the end that pattern/retention can be looked up with only one index.
Maybe it would be good to have some comment that illustrates how that works, I imagine the `s.index` slice to look like a partitioned list of retentions where each partition is a pattern:
```
|-----------------------------------------------------------------------------|
|   pattern1          |  pattern2   |            pattern3                     |
|-----------------------------------------------------------------------------|
| ret1 | ret2 | ret3 | ret4 | ret5 | ret6 | ret7 | ret8 | ret9 | ret10 |
|-----------------------------------------------------------------------------|

1) if metric matches patternX -> find the best retention within patternX
2) otherwise -> skip len(patternX.retentions) and back to 1)
```